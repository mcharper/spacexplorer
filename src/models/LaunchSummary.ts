export interface LaunchSummary {
  rocketName: string;
  launchDateUtc: string;
  nationality: string;
}
