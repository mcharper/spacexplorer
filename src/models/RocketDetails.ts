export interface RocketDetails {
  id: string;
  name: string;
  type: string;
  description: string;
  images: string[];
  stages: number;
  costPerLaunch: number;
  country: string;
  height: number;
  diameter: number;
}
