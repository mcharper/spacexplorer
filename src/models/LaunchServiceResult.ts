import { LaunchSummary } from "./LaunchSummary";

export interface LaunchServiceResult {
  isLoading: boolean;
  launchList: LaunchSummary[];
}
