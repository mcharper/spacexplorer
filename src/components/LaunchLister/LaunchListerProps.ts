import { LaunchSummary } from "../../models/LaunchSummary";

export interface LaunchListerProps {
  launchList: LaunchSummary[];
}
