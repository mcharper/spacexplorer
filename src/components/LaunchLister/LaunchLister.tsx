import { AgGridReact, AgGridColumn } from "ag-grid-react";

import { LaunchListerProps } from "./LaunchListerProps";

import "ag-grid-community/dist/styles/ag-grid.css";
import "ag-grid-community/dist/styles/ag-theme-alpine.css";
import { useLaunchListerStyles } from "./hooks/useLaunchListerStyles";

export const LaunchLister = (props: LaunchListerProps): JSX.Element => {
  const classes = useLaunchListerStyles();

  return (
    <div>
      <h2>Launch List</h2>

      <div className={classes.launchList + " ag-theme-alpine"}>
        <AgGridReact rowData={props.launchList}>
          <AgGridColumn
            field="missionName"
            sortable={true}
            filter={true}
          ></AgGridColumn>

          <AgGridColumn
            field="launchDateUtc"
            sortable={true}
            cellRenderer={(params) =>
              params.value ? new Date(params.value).toLocaleDateString() : ""
            }
          ></AgGridColumn>

          <AgGridColumn
            field="rocketId"
            onCellClicked={(e) => {
              props.setCurrentRocketId(e.value);
            }}
          ></AgGridColumn>
        </AgGridReact>
      </div>
    </div>
  );
};
