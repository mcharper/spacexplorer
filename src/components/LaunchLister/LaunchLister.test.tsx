import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import { LaunchLister } from "./LaunchLister";
import { LaunchSummary } from "../../models/LaunchSummary";

let launchList: LaunchSummary[];

beforeAll(() => {
  launchList = [
    {
      missionName: "Mission-X",
      launchDateUtc: "2017-06-23T19:10:000Z",
      rocketId: "falcon1",
    },
    {
      missionName: "Mission-Y",
      launchDateUtc: "2017-06-24T19:10:000Z",
      rocketId: "falcon1",
    },
    {
      missionName: "Mission-Z",
      launchDateUtc: "2017-06-25T19:10:000Z",
      rocketId: "falcon1",
    }
  ];
});

test("renders a grid to display the list", () => {
  render(
    <LaunchLister launchList={launchList} setCurrentRocketId={jest.fn()} />
  );

  const grid = screen.getByRole("grid");
  expect(grid).toBeInTheDocument();
});

test("renders a header for each column represented in the data", () => {
  render(
    <LaunchLister launchList={launchList} setCurrentRocketId={jest.fn()} />
  );
  const columnHeaders = screen.getAllByRole("columnheader");
  expect(columnHeaders).toHaveLength(3);
});

test("renders right text for each column header", () => {
  render(
    <LaunchLister launchList={launchList} setCurrentRocketId={jest.fn()} />
  );

  const columnHeaders = screen.getAllByRole("columnheader");

  expect(within(columnHeaders[0]).getByText(/Mission Name/)).not.toBeNull();
  expect(within(columnHeaders[1]).getByText(/Launch Date Utc/)).not.toBeNull();
  expect(within(columnHeaders[2]).getByText(/Rocket Id/)).not.toBeNull();
});

test("renders grid cells to display the data", () => {
  // agGrid creates elements with the role "gridcell"
  // for each column, for each row of data
  const expectedGridCellCount = launchList.length * 3;

  render(
    <LaunchLister launchList={launchList} setCurrentRocketId={jest.fn()} />
  );

  const grid = screen.getByRole("grid");
  const gridCellCount = within(grid).getAllByRole("gridcell");
  expect(gridCellCount).toHaveLength(expectedGridCellCount);
});

test("(for illustration purposes) the data is shown in the original order and each row is assigned an ascending index", () => {
  render(
    <LaunchLister launchList={launchList} setCurrentRocketId={jest.fn()} />
  );

  const valuesInColumn = screen.queryAllByText(/Mission-/);

  expect(valuesInColumn[0].textContent).toBe("Mission-X"); 
  expect(valuesInColumn[1].textContent).toBe("Mission-Y");
  expect(valuesInColumn[2].textContent).toBe("Mission-Z");

  expect(valuesInColumn[0].parentNode.getAttribute("row-index")).toBe("0");
  expect(valuesInColumn[1].parentNode.getAttribute("row-index")).toBe("1");
  expect(valuesInColumn[2].parentNode.getAttribute("row-index")).toBe("2");

});

test("clicking once on the mission name column header sorts the data in ascending alphabetical order", () => {
  // agGrid doesn't move the elements in the DOM when sorting
  // So they are in the order shown in the mock
  // It just changes the row indices
  render(
    <LaunchLister launchList={launchList} setCurrentRocketId={jest.fn()} />
  );

  const missionNameHeader = screen.getByText("Mission Name");
  userEvent.click(missionNameHeader);

  const valuesInColumn = screen.queryAllByText(/Mission-/);

  // (For illustration purposes) agGrid doesn't actually reorder the DOM elements
  expect(valuesInColumn[0].textContent).toBe("Mission-X");
  expect(valuesInColumn[1].textContent).toBe("Mission-Y");
  expect(valuesInColumn[2].textContent).toBe("Mission-Z");
  
  // It just moves the row indices around
  expect(valuesInColumn[0].parentNode.getAttribute("row-index")).toBe("0");
  expect(valuesInColumn[1].parentNode.getAttribute("row-index")).toBe("1");
  expect(valuesInColumn[2].parentNode.getAttribute("row-index")).toBe("2");
});

test("clicking twice on the mission name column header sorts the data in descending alphabetical order", () => {
  render(
    <LaunchLister launchList={launchList} setCurrentRocketId={jest.fn()} />
  );

  const missionNameHeader = screen.getByText("Mission Name");
  userEvent.click(missionNameHeader);
  userEvent.click(missionNameHeader);

  const valuesInColumn = screen.queryAllByText(/Mission-/);

  // (For illustration purposes) agGrid doesn't actually reorder the DOM elements
  expect(valuesInColumn[0].textContent).toBe("Mission-X");
  expect(valuesInColumn[1].textContent).toBe("Mission-Y");
  expect(valuesInColumn[2].textContent).toBe("Mission-Z");
  
  // It just moves the row indices around
  expect(valuesInColumn[0].parentNode.getAttribute("row-index")).toBe("2");
  expect(valuesInColumn[1].parentNode.getAttribute("row-index")).toBe("1");
  expect(valuesInColumn[2].parentNode.getAttribute("row-index")).toBe("0");
});

test("clicking three times on the mission name column header restores the original sort order", () => {
  render(
    <LaunchLister launchList={launchList} setCurrentRocketId={jest.fn()} />
  );

  const missionNameHeader = screen.getByText("Mission Name");
  userEvent.click(missionNameHeader);
  userEvent.click(missionNameHeader);
  userEvent.click(missionNameHeader);

  const valuesInColumn = screen.queryAllByText(/Mission-/);

  // (For illustration purposes) agGrid doesn't actually reorder the DOM elements
  expect(valuesInColumn[0].textContent).toBe("Mission-X");
  expect(valuesInColumn[1].textContent).toBe("Mission-Y");
  expect(valuesInColumn[2].textContent).toBe("Mission-Z");
  
  // It just moves the row indices around
  expect(valuesInColumn[0].parentNode.getAttribute("row-index")).toBe("0");
  expect(valuesInColumn[1].parentNode.getAttribute("row-index")).toBe("1");
  expect(valuesInColumn[2].parentNode.getAttribute("row-index")).toBe("2");
});
