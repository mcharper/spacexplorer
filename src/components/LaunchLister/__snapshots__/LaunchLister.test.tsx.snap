// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a grid that matches the expected snapshot 1`] = `
<div
  aria-colcount="3"
  aria-rowcount="4"
  class="ag-root ag-unselectable ag-layout-normal"
  ref="gridBody"
  role="grid"
  unselectable="on"
>
  
        
  <!--AG-HEADER-ROOT-->
  <div
    class="ag-header ag-focus-managed ag-pivot-off"
    ref="gridHeader"
    role="presentation"
    style="height: 49px; min-height: 49px;"
    unselectable="on"
  >
    <div
      class="ag-pinned-left-header ag-hidden"
      role="presentation"
      style="width: 0px; max-width: 0px; min-width: 0px;"
    >
      <div
        aria-rowindex="1"
        class="ag-header-row ag-header-row-column"
        role="row"
        style="top: 0px; height: 48px; width: 0px;"
      />
    </div>
    <div
      class="ag-header-viewport"
      role="presentation"
    >
      
            
      <div
        class="ag-header-container"
        ref="eCenterContainer"
        role="rowgroup"
        style="width: 600px;"
      >
        <div
          aria-rowindex="1"
          class="ag-header-row ag-header-row-column"
          role="row"
          style="top: 0px; height: 48px; width: 600px;"
        >
          <div
            aria-colindex="1"
            aria-sort="none"
            class="ag-header-cell ag-header-cell-sortable ag-focus-managed"
            col-id="missionName"
            role="columnheader"
            style="width: 200px; left: 0px;"
            tabindex="-1"
            unselectable="on"
          >
            
            
            <div
              class="ag-header-cell-resize ag-hidden"
              ref="eResize"
              role="presentation"
            />
            <div
              class="ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-header-select-all ag-hidden"
              role="presentation"
            >
              
                
              <div
                class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                ref="eLabel"
                role="presentation"
              />
              
                
              <div
                class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                ref="eWrapper"
                role="presentation"
              >
                
                    
                <input
                  aria-label="Press Space to toggle all rows selection (unchecked)"
                  class="ag-input-field-input ag-checkbox-input"
                  id="ag-76-input"
                  ref="eInput"
                  tabindex="-1"
                  type="checkbox"
                />
                
                
              </div>
              
            
            </div>
            
        
            <div
              class="ag-cell-label-container ag-header-cell-sorted-none"
              role="presentation"
            >
              
            
              <span
                aria-hidden="true"
                class="ag-header-icon ag-header-cell-menu-button"
                ref="eMenu"
              >
                <span
                  class="ag-icon ag-icon-menu"
                  role="presentation"
                  unselectable="on"
                />
              </span>
              
            
              <div
                class="ag-header-cell-label"
                ref="eLabel"
                role="presentation"
                unselectable="on"
              >
                
                
                <span
                  class="ag-header-cell-text"
                  ref="eText"
                  unselectable="on"
                >
                  Mission Name
                </span>
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                  ref="eFilter"
                >
                  <span
                    class="ag-icon ag-icon-filter"
                    role="presentation"
                    unselectable="on"
                  />
                </span>
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-sort-order ag-hidden"
                  ref="eSortOrder"
                />
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-sort-ascending-icon ag-hidden"
                  ref="eSortAsc"
                >
                  <span
                    class="ag-icon ag-icon-asc"
                    role="presentation"
                    unselectable="on"
                  />
                </span>
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-sort-descending-icon ag-hidden"
                  ref="eSortDesc"
                >
                  <span
                    class="ag-icon ag-icon-desc"
                    role="presentation"
                    unselectable="on"
                  />
                </span>
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-sort-none-icon ag-hidden"
                  ref="eSortNone"
                >
                  <span
                    class="ag-icon ag-icon-none"
                    role="presentation"
                    unselectable="on"
                  />
                </span>
                
            
              </div>
              
        
            </div>
          </div>
          <div
            aria-colindex="2"
            aria-sort="none"
            class="ag-header-cell ag-header-cell-sortable ag-focus-managed"
            col-id="launchDateUtc"
            role="columnheader"
            style="width: 200px; left: 200px;"
            tabindex="-1"
            unselectable="on"
          >
            
            
            <div
              class="ag-header-cell-resize ag-hidden"
              ref="eResize"
              role="presentation"
            />
            <div
              class="ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-header-select-all ag-hidden"
              role="presentation"
            >
              
                
              <div
                class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                ref="eLabel"
                role="presentation"
              />
              
                
              <div
                class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                ref="eWrapper"
                role="presentation"
              >
                
                    
                <input
                  aria-label="Press Space to toggle all rows selection (unchecked)"
                  class="ag-input-field-input ag-checkbox-input"
                  id="ag-79-input"
                  ref="eInput"
                  tabindex="-1"
                  type="checkbox"
                />
                
                
              </div>
              
            
            </div>
            
        
            <div
              class="ag-cell-label-container ag-header-cell-sorted-none"
              role="presentation"
            >
              
            
              
            
              <div
                class="ag-header-cell-label"
                ref="eLabel"
                role="presentation"
                unselectable="on"
              >
                
                
                <span
                  class="ag-header-cell-text"
                  ref="eText"
                  unselectable="on"
                >
                  Launch Date Utc
                </span>
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                  ref="eFilter"
                >
                  <span
                    class="ag-icon ag-icon-filter"
                    role="presentation"
                    unselectable="on"
                  />
                </span>
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-sort-order ag-hidden"
                  ref="eSortOrder"
                />
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-sort-ascending-icon ag-hidden"
                  ref="eSortAsc"
                >
                  <span
                    class="ag-icon ag-icon-asc"
                    role="presentation"
                    unselectable="on"
                  />
                </span>
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-sort-descending-icon ag-hidden"
                  ref="eSortDesc"
                >
                  <span
                    class="ag-icon ag-icon-desc"
                    role="presentation"
                    unselectable="on"
                  />
                </span>
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-sort-none-icon ag-hidden"
                  ref="eSortNone"
                >
                  <span
                    class="ag-icon ag-icon-none"
                    role="presentation"
                    unselectable="on"
                  />
                </span>
                
            
              </div>
              
        
            </div>
          </div>
          <div
            aria-colindex="3"
            class="ag-header-cell ag-focus-managed"
            col-id="rocketId"
            role="columnheader"
            style="width: 200px; left: 400px;"
            tabindex="-1"
            unselectable="on"
          >
            
            
            <div
              class="ag-header-cell-resize ag-hidden"
              ref="eResize"
              role="presentation"
            />
            <div
              class="ag-labeled ag-label-align-right ag-checkbox ag-input-field ag-header-select-all ag-hidden"
              role="presentation"
            >
              
                
              <div
                class="ag-input-field-label ag-label ag-hidden ag-checkbox-label"
                ref="eLabel"
                role="presentation"
              />
              
                
              <div
                class="ag-wrapper ag-input-wrapper ag-checkbox-input-wrapper"
                ref="eWrapper"
                role="presentation"
              >
                
                    
                <input
                  aria-label="Press Space to toggle all rows selection (unchecked)"
                  class="ag-input-field-input ag-checkbox-input"
                  id="ag-82-input"
                  ref="eInput"
                  tabindex="-1"
                  type="checkbox"
                />
                
                
              </div>
              
            
            </div>
            
        
            <div
              class="ag-cell-label-container"
              role="presentation"
            >
              
            
              
            
              <div
                class="ag-header-cell-label"
                ref="eLabel"
                role="presentation"
                unselectable="on"
              >
                
                
                <span
                  class="ag-header-cell-text"
                  ref="eText"
                  unselectable="on"
                >
                  Rocket Id
                </span>
                
                
                <span
                  aria-hidden="true"
                  class="ag-header-icon ag-header-label-icon ag-filter-icon ag-hidden"
                  ref="eFilter"
                >
                  <span
                    class="ag-icon ag-icon-filter"
                    role="presentation"
                    unselectable="on"
                  />
                </span>
                
                
                
                
                
                
                
                
                
            
              </div>
              
        
            </div>
          </div>
        </div>
      </div>
      
        
    </div>
    <div
      class="ag-pinned-right-header ag-hidden"
      role="presentation"
      style="width: 0px; max-width: 0px; min-width: 0px;"
    >
      <div
        aria-rowindex="1"
        class="ag-header-row ag-header-row-column"
        role="row"
        style="top: 0px; height: 48px; width: 0px;"
      />
    </div>
  </div>
  
        
  <div
    class="ag-floating-top"
    ref="eTop"
    role="presentation"
    style="min-height: 0px; height: 0px; display: none; overflow-y: hidden;"
    unselectable="on"
  >
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-pinned-left-floating-top ag-hidden"
      name="topLeft"
      ref="topLeftContainer"
      role="presentation"
      unselectable="on"
    />
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-floating-top-viewport"
      name="topCenter"
      ref="topCenterContainer"
      role="presentation"
      unselectable="on"
    >
      
                
      <div
        class="ag-floating-top-container"
        ref="eContainer"
        role="presentation"
        style="width: 600px;"
        unselectable="on"
      />
      
            
    </div>
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-pinned-right-floating-top ag-hidden"
      name="topRight"
      ref="topRightContainer"
      role="presentation"
      unselectable="on"
    />
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-floating-top-full-width-container"
      name="topFullWidth"
      ref="topFullWidthContainer"
      role="presentation"
      unselectable="on"
    />
    
        
  </div>
  
        
  <div
    class="ag-body-viewport ag-layout-normal ag-row-no-animation"
    ref="eBodyViewport"
    role="presentation"
  >
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-pinned-left-cols-container ag-hidden"
      name="left"
      ref="leftContainer"
      role="presentation"
      style="height: 126px;"
      unselectable="on"
    >
      <div
        aria-rowindex="2"
        class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-first"
        comp-id="84"
        role="row"
        row-id="0"
        row-index="0"
        style="transform: translateY(0px); height: 42px;"
      />
      <div
        aria-rowindex="3"
        class="ag-row-odd ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute"
        comp-id="85"
        role="row"
        row-id="1"
        row-index="1"
        style="transform: translateY(42px); height: 42px;"
      />
      <div
        aria-rowindex="4"
        class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-last"
        comp-id="86"
        role="row"
        row-id="2"
        row-index="2"
        style="transform: translateY(84px); height: 42px;"
      />
    </div>
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-center-cols-clipper"
      name="center"
      ref="centerContainer"
      role="presentation"
      style="height: 126px;"
      unselectable="on"
    >
      
                
      <div
        class="ag-center-cols-viewport"
        ref="eViewport"
        role="presentation"
        style="height: 100%;"
      >
        
                    
        <div
          class="ag-center-cols-container"
          ref="eContainer"
          role="rowgroup"
          style="width: 600px; height: 126px;"
          unselectable="on"
        >
          <div
            aria-rowindex="2"
            class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-first"
            comp-id="87"
            role="row"
            row-id="0"
            row-index="0"
            style="transform: translateY(0px); height: 42px;"
          >
            <div
              aria-colindex="1"
              class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
              col-id="missionName"
              comp-id="91"
              role="gridcell"
              style="left: 0px; width: 200px;"
              tabindex="-1"
              unselectable="on"
            >
              Mission-X
            </div>
            <div
              aria-colindex="2"
              class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
              col-id="launchDateUtc"
              comp-id="92"
              role="gridcell"
              style="left: 200px; width: 200px;"
              tabindex="-1"
              unselectable="on"
            />
            <div
              aria-colindex="3"
              class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
              col-id="rocketId"
              comp-id="93"
              role="gridcell"
              style="left: 400px; width: 200px;"
              tabindex="-1"
              unselectable="on"
            >
              falcon1
            </div>
          </div>
          <div
            aria-rowindex="3"
            class="ag-row-odd ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute"
            comp-id="88"
            role="row"
            row-id="1"
            row-index="1"
            style="transform: translateY(42px); height: 42px;"
          >
            <div
              aria-colindex="1"
              class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
              col-id="missionName"
              comp-id="95"
              role="gridcell"
              style="left: 0px; width: 200px;"
              tabindex="-1"
              unselectable="on"
            >
              Mission-Y
            </div>
            <div
              aria-colindex="2"
              class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
              col-id="launchDateUtc"
              comp-id="96"
              role="gridcell"
              style="left: 200px; width: 200px;"
              tabindex="-1"
              unselectable="on"
            />
            <div
              aria-colindex="3"
              class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
              col-id="rocketId"
              comp-id="97"
              role="gridcell"
              style="left: 400px; width: 200px;"
              tabindex="-1"
              unselectable="on"
            >
              falcon1
            </div>
          </div>
          <div
            aria-rowindex="4"
            class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-last"
            comp-id="89"
            role="row"
            row-id="2"
            row-index="2"
            style="transform: translateY(84px); height: 42px;"
          >
            <div
              aria-colindex="1"
              class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
              col-id="missionName"
              comp-id="99"
              role="gridcell"
              style="left: 0px; width: 200px;"
              tabindex="-1"
              unselectable="on"
            >
              Mission-Z
            </div>
            <div
              aria-colindex="2"
              class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
              col-id="launchDateUtc"
              comp-id="100"
              role="gridcell"
              style="left: 200px; width: 200px;"
              tabindex="-1"
              unselectable="on"
            />
            <div
              aria-colindex="3"
              class="ag-cell ag-cell-not-inline-editing ag-cell-normal-height ag-cell-value"
              col-id="rocketId"
              comp-id="101"
              role="gridcell"
              style="left: 400px; width: 200px;"
              tabindex="-1"
              unselectable="on"
            >
              falcon1
            </div>
          </div>
        </div>
        
                
      </div>
      
            
    </div>
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-pinned-right-cols-container ag-hidden"
      name="right"
      ref="rightContainer"
      role="presentation"
      style="height: 126px;"
      unselectable="on"
    >
      <div
        aria-rowindex="2"
        class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-first"
        comp-id="90"
        role="row"
        row-id="0"
        row-index="0"
        style="transform: translateY(0px); height: 42px;"
      />
      <div
        aria-rowindex="3"
        class="ag-row-odd ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute"
        comp-id="94"
        role="row"
        row-id="1"
        row-index="1"
        style="transform: translateY(42px); height: 42px;"
      />
      <div
        aria-rowindex="4"
        class="ag-row-even ag-row-no-focus ag-row ag-row-level-0 ag-row-position-absolute ag-row-last"
        comp-id="98"
        role="row"
        row-id="2"
        row-index="2"
        style="transform: translateY(84px); height: 42px;"
      />
    </div>
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-full-width-container"
      name="fullWidth"
      ref="fullWidthContainer"
      role="presentation"
      style="height: 126px;"
      unselectable="on"
    />
    
        
  </div>
  
        
  <div
    class="ag-floating-bottom"
    ref="eBottom"
    role="presentation"
    style="min-height: 0px; height: 0px; display: none; overflow-y: hidden;"
    unselectable="on"
  >
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-pinned-left-floating-bottom ag-hidden"
      name="bottomLeft"
      ref="bottomLeftContainer"
      role="presentation"
      unselectable="on"
    />
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-floating-bottom-viewport"
      name="bottomCenter"
      ref="bottomCenterContainer"
      role="presentation"
      unselectable="on"
    >
      
                
      <div
        class="ag-floating-bottom-container"
        ref="eContainer"
        role="presentation"
        style="width: 600px;"
        unselectable="on"
      />
      
            
    </div>
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-pinned-right-floating-bottom ag-hidden"
      name="bottomRight"
      ref="bottomRightContainer"
      role="presentation"
      unselectable="on"
    />
    
            
    <!--AG-ROW-CONTAINER-->
    <div
      class="ag-floating-bottom-full-width-container"
      name="bottomFullWidth"
      ref="bottomFullWidthContainer"
      role="presentation"
      unselectable="on"
    />
    
        
  </div>
  
        
  <!--AG-FAKE-HORIZONTAL-SCROLL-->
  <div
    aria-hidden="true"
    class="ag-body-horizontal-scroll"
    style="height: 0px; max-height: 0px; min-height: 0px;"
  >
    
            
    <div
      class="ag-horizontal-left-spacer"
      ref="eLeftSpacer"
      style="width: 0px; max-width: 0px; min-width: 0px;"
    />
    
            
    <div
      class="ag-body-horizontal-scroll-viewport"
      ref="eViewport"
      style="height: 0px; max-height: 0px; min-height: 0px;"
    >
      
                
      <div
        class="ag-body-horizontal-scroll-container"
        ref="eContainer"
        style="height: 0px; max-height: 0px; min-height: 0px; width: 600px;"
      />
      
            
    </div>
    
            
    <div
      class="ag-horizontal-right-spacer"
      ref="eRightSpacer"
      style="width: 0px; max-width: 0px; min-width: 0px;"
    />
    
        
  </div>
  
        
  <!--AG-OVERLAY-WRAPPER-->
  <div
    aria-hidden="true"
    class="ag-overlay ag-hidden"
  >
    
            
    <div
      class="ag-overlay-panel"
    >
      
                
      <div
        class="ag-overlay-wrapper ag-layout-normal"
        ref="eOverlayWrapper"
      />
      
            
    </div>
    
        
  </div>
  
    
</div>
`;
